File: client/src/types/spreadsheet.ts
// Represents data required to create a spreadsheet
export type SpreadsheetInit = {
  title: string;
  description: string;
  deadline: string;
  colTitles: string[];
};

export type SpreadsheetInitErr = Omit<Partial<SpreadsheetInit>, "colTitles"> & {
  // error key is a non field error
  error?: string;
  colTitles?: string;
};

// Represents a spreadsheet retrieved from the api
export type Spreadsheet = {
  id: string;
  title: string;
  description: string;
  owner: string;
  createdAt: string;
  updatedAt: string;
  data: string[][];
  deadline: string;
};

File: api/models/spreadsheet.go
package models

import "time"

// SpreadsheetInit represents the payload required to create a new spreadsheet.
type SpreadsheetInit struct {
	Title       string    `json:"title" binding:"required,max=255"`
	Description string    `json:"description" binding:"required"`
	Deadline    time.Time `json:"deadline" time_format:"2006-01-02T15:04:05Z07:00" binding:"required"` // Deadline in RFC3339 format
	ColTitles   []string  `json:"colTitles" binding:"required,min=1"`
}

// Spreadsheet represents a spreadsheet stored in the database.
type Spreadsheet struct {
	ID          string    `json:"id"`
	Title       string    `json:"title"`
	Description string    `json:"description"`
	Owner       string    `json:"owner"`
	CreatedAt   time.Time `json:"createdAt"`
	UpdatedAt   time.Time `json:"updatedAt"`
	Data        []byte    `json:"data"` // [][]string stored as jsonb
	Deadline    time.Time `json:"deadline"`
}
